tests/test_eqvar.py

```diff
@@ -0,0 +1,207 @@
+            tf.debugging.assert_near(energy1, energy2)
+        
+    def test_pinet2(self, mocked_data):
+        pinet = PiNet2(
+            atom_types=[0, 1],
+        )
+
+        for batch in mocked_data:
+            batch1 = batch.copy()
+            energy1 = pinet(batch1)
+
+            batch2 = batch.copy()
+            batch2['coord'] = rotate(batch2['coord'], 42.)
+            energy2 = pinet(batch2)
+            tf.debugging.assert_near(energy1, energy2)
+
+    def test_pinet2_p5_dot(self, mocked_data):
+        pinet = PiNet2P5Dot(
+            atom_types=[0, 1],
+        )
+
+        for batch in mocked_data:
+            batch1 = batch.copy()
+            energy1 = pinet(batch1)
+
+            batch2 = batch.copy()
+            batch2['coord'] = rotate(batch2['coord'], 42.)
+            energy2 = pinet(batch2)
+            np.testing.assert_allclose(energy1, energy2, rtol=1e-4, atol=1e-4)
+
+    def test_pinet2_p5_prod(self, mocked_data):
+        pinet = PiNet2P5Prod(
+            atom_types=[0, 1],
+        )
+
+        for batch in mocked_data:
+            batch1 = batch.copy()
+            energy1 = pinet(batch1)
+
+            batch2 = batch.copy()
+            batch2['coord'] = rotate(batch2['coord'], 42.)
+            energy2 = pinet(batch2)
+            np.testing.assert_allclose(energy1, energy2, rtol=1e-3, atol=1e-3)
+
+    def test_pinet2_p5_irrep_dot(self, mocked_data):
+        pinet = PiNet2P5Dot(
+            atom_types=[0, 1],
+        )
+        
+        for batch in mocked_data:
+            batch1 = batch.copy()
+            energy1 = pinet(batch1)
+
+            batch2 = batch.copy()
+            batch2['coord'] = rotate(tf.cast(batch2['coord'], tf.float32), random.randint(0, 4123))
+            energy2 = pinet(batch2)
+            np.testing.assert_allclose(energy1, energy2, rtol=1e-4, atol=1e-4)
+
+    def test_pinet2_p5_irrep_prod(self, mocked_data):
+        pinet = PiNet2P5Prod(
+            atom_types=[0, 1],
+        )
+        
+        for batch in mocked_data:
+            batch1 = batch.copy()
+            energy1 = pinet(batch1)
+
+            batch2 = batch.copy()
+            batch2['coord'] = rotate(tf.cast(batch2['coord'], tf.float32), random.randint(0, 4123))
+            energy2 = pinet(batch2)
+            np.testing.assert_allclose(energy1, energy2, rtol=1e-4, atol=1e-4)
+
+    def test_pinet2_p5_irrep_combine(self, mocked_data):
+        pinet = PiNet2P5Combine(
+            atom_types=[0, 1],
+        )
+        
+        for batch in mocked_data:
+            batch1 = batch.copy()
+            energy1 = pinet(batch1)
+
+            batch2 = batch.copy()
+            batch2['coord'] = rotate(tf.cast(batch2['coord'], tf.float32), random.randint(0, 4123))
+            energy2 = pinet(batch2)
+            np.testing.assert_allclose(energy1, energy2, rtol=1e-4, atol=1e-4)
+
+
+    def test_pinet2_p5_irrep_prod_internal(self, mocked_data):
+        from pinn.layers import (
```

Comment:
Is this meant to test the p3 variable after the whole PiNet? I don't think it's necessary to replicate the whole code here. `pinet2_p5_prod(tensors); tensors['p3']` should do.

Reply:
Can't. `Preprocessing` copies the tensor, and layers won't change input tensors; I can't find a way to expose intermediate variables from the network.

Reply:
Ah, right, that was to prevent layers from mutating, maybe give the network a `_tensors` attribute to access the intemediate dict of tensors?

Reply:
Lisanne wants to return internal variables and transform them into desired channels. I try to return an extra variable, but it looks super ugly https://github.com/Teoroo-CMC/PiNN/pull/18#discussion_r1644197428.